cmd/util/ledger/migrations/multiple_contract_migration.go

```diff
@@ -0,0 +1,261 @@
+		logger.Info().
+			Err(err).
+			Str("address", flow.BytesToAddress([]byte(registerId.Owner)).HexWithPrefix()).
+			Msg("Using exceptional migration for address")
+		return em(p, logger)
+	}
+
+	return migrateValue(p, logger)
+}
+
+func migrateValue(p ledger.Payload, l zerolog.Logger) ([]ledger.Payload, error) {
+
+	// we don't need the the empty code register
+	value := p.Value
+	address := flow.BytesToAddress(p.Key.KeyParts[0].Value)
+	if len(value) == 0 {
+		return []ledger.Payload{}, nil
+	}
+
+	code := string(value)
+	program, err := parser2.ParseProgram(code)
+	if err != nil {
+		l.Error().
+			Err(err).
+			Str("address", address.Hex()).
+			Str("code", code).
+			Msg("Cannot parse program at address")
+		return nil, err
+	}
+	declarations := program.Declarations
+
+	// find import declarations
+	importDeclarations := make([]ast.Declaration, 0)
+	for _, d := range declarations {
+		if _, isImport := d.(*ast.ImportDeclaration); isImport {
+			importDeclarations = append(importDeclarations, d)
+		}
+	}
+
+	// sort them backwards because we will be slicing the code string
+	sort.SliceStable(importDeclarations, func(i, j int) bool {
+		return importDeclarations[i].StartPosition().Offset > importDeclarations[j].StartPosition().Offset
+	})
+
+	imports := ""
+	for _, d := range importDeclarations {
+		// aggregate imports for later use
+		imports = imports + code[d.StartPosition().Offset:d.EndPosition().Offset+1] + "\n"
+		// remove imports from the code
+		code = code[:d.StartPosition().Offset] + code[d.EndPosition().Offset+1:]
+	}
+
+	// parse the code again to get accurate locations of the remaining declarations
+	program, err = parser2.ParseProgram(code)
+	if err != nil {
+		l.Error().
+			Err(err).
+			Str("address", address.Hex()).
+			Str("code", code).
+			Msg("Cannot parse program at address after removing declarations")
+		return nil, err
+	}
+	declarations = program.Declarations
+
+	if len(declarations) == 0 {
+		// If there is no declarations. Only comments? was this legal before?
+		// error just in case
+		// alternative would be removing the register
+		l.Error().
+			Str("address", address.Hex()).
+			Str("code", code).
+			Msg("No declarations at address")
+		return []ledger.Payload{}, fmt.Errorf("no declarations at address %s", address.Hex())
+	}
+
+	if len(declarations) > 2 {
+		l.Error().
+			Str("address", address.Hex()).
+			Str("code", code).
+			Msg("More than two declarations at address")
+		return []ledger.Payload{}, fmt.Errorf("more than two declarations at address %s", address.Hex())
+	}
+
+	if len(declarations) == 1 {
+		// If there is one declaration move it to the new key
+		l.Debug().
+			Str("address", address.Hex()).
+			Msg("Single contract or interface at address moved to new key")
+		p.Key = addNameToKey(p.Key, declarations[0].DeclarationIdentifier().Identifier)
+		return []ledger.Payload{p}, nil
+	}
+
+	// We have two declarations. Due to the current rules one of them is an interface and one is a contract.
```

Comment:
Currently zero or one contract, and zero or more contract interfaces are allowed. Was the idea here that we only support the case where we have a contract and a contract interface, because that's the only case in the real world (on the current testnet and mainnet)?

Reply:
Thats correct. The migration only supports the cases where there is:
- no contracts/interfaces
- one contract/interface (it will never be one interface as that is currently not allowed)
- one contract and one interface